### PR TITLE
tests: Run tagged tests with `--changed` when shared contexts change

### PR DESCRIPTION
**Pull request by issyl0**
-----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them? Performance claims (e.g. "this is faster") must include [Hyperfine](https://github.com/sharkdp/hyperfine) benchmarks.
- [x] Have you written new tests (excluding integration tests) for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew lgtm` (style, typechecking and tests) with your changes locally?

-----

- [x] AI was used to generate or assist with generating this PR.

- Assisted by GPT-5.3 Codex (Medium reasoning). Human review, tests and tweaks.

-----

- When a shared context file changes, the tests that include that shared context should be run when `brew tests --changed` is invoked. This ensures that relevant test failures are caught.
- I found this bug while working on #22469 where I changed shared contexts, ran `brew lgtm` which ran `brew tests --changed` which lulled me into a false sense of security that the changes were fine, until CI ran the full test suite[1].
- Add some standard behaviour tests for `brew tests --changed` as well as there was zero coverage before.

[1]: https://github.com/Homebrew/brew/pull/22469#issuecomment-4583056266

### CODE REVIEW

File: Library/Homebrew/dev-cmd/tests.rb [hunk 1]
```diff
@@ -241,19 +241,60 @@ def changed_test_files
...
+      def file_uses_rspec_tag?(path, tag)
+        escaped_tag = Regexp.escape(tag)
+        rspec_declaration_methods = %w[describe context it specify example].join("|")
+        rspec_declaration_regex = /^\s*(?:RSpec\.)?(?:#{rspec_declaration_methods})\b/
+        symbol_tag_regex = /(?:^|[,(])\s*:#{escaped_tag}\b/
+        hash_tag_regex = /(?:^|[,(])\s*#{escaped_tag}:\s*(?:true|false|nil|:[a-z_]\w*|[a-z_]\w*)?/i
+
+        path.read.each_line.any? do |line|
+          line.match?(rspec_declaration_regex) &&
+            (line.match?(symbol_tag_regex) || line.match?(hash_tag_regex))
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb]: This is a little opaque, some comments could help.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks! One comment, don't need rereview after.
